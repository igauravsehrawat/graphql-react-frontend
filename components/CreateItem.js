import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Mutation } from 'react-apollo';
import Router from 'next/router';
import Form from './styles/Form';
import Error from './ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    price: 0,
    image: '',
    largeImage: '',
  };

  handleChange = e => {
    const { name, type, value } = e.target;
    // console.log({ name, value, type });
    const typeValue = type === 'number' ? parseFloat(value) : value;
    this.setState({
      [name]: typeValue,
    });
  };

  handleUpload = async e => {
    const { files } = e.target;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'devFits');
    const res = await fetch(
      'https://api.cloudinary.com/v1_1/af0lqrya/image/upload',
      {
        method: 'POST',
        body: data,
      }
    );
    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    });
  };

  render() {
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
            onSubmit={async e => {
              // prevent default
              e.preventDefault();
              // call mutation
              const res = await createItem();
              // console.log({ res });
              // change the route
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              });
            }}
          >
            <Error error={error} />
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="image">Image</label>
              <input
                id="image"
                name="image"
                type="file"
                onChange={this.handleUpload}
              />
              {this.state.image && (
                <img width="200" src={this.state.image} alt="Upload preview" />
              )}
              <label htmlFor="title">Title</label>
              <input
                id="title"
                name="title"
                type="text"
                value={this.state.title}
                placeholder="Enter title"
                required
                onChange={this.handleChange}
              />
              <label htmlFor="price">Price</label>
              <input
                id="price"
                name="price"
                type="number"
                value={this.state.price}
                placeholder="Enter price"
                required
                onChange={this.handleChange}
              />
              <label htmlFor="description"></label>
              <textarea
                id="description"
                name="description"
                value={this.state.description}
                placeholder="Enter description"
                required
                onChange={this.handleChange}
              />
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };
