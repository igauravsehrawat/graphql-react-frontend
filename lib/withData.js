import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint } from '../config';
import { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION } from '../components/Cart';

function createClient({ headers }) {
  return new ApolloClient({
    // ? true: false, i know
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_root, variables, { cache }) {
            const { cartOpen } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            console.log('data', cartOpen);
            const newData = {
              data: {
                cartOpen: !cartOpen,
              }
            }
            cache.writeData(newData);
            return newData;
          }
        },
      },
      defaults: {
        cartOpen: true,
      }
    }
  });
}

export default withApollo(createClient);
