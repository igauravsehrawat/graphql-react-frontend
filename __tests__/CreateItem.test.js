import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import { MockedProvider } from 'react-apollo/test-utils';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';
import { fakeItem } from '../lib/testUtils';

// mock the global fetch API
// fetch is global browser API
const dogImage = 'https://dog.com/dog.jpg';
global.fetch = jest.fn().mockResolvedValue({
  json: () => ({
    secure_url: dogImage,
    eager: [
      {
        secure_url: dogImage,
      },
    ],
  }),
});

describe('<CreateItem />', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );
    const form = wrapper.find('form');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it('uploads a file when changed', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );
    const input = wrapper.find('input[type="file"]');
    input.simulate('change', { target: { files: ['fakedog.jpg'] } });
    await wait();
    const component = wrapper.find('CreateItem').instance();
    expect(component.state.image).toEqual(dogImage);
    expect(component.state.largeImage).toEqual(dogImage);
    expect(global.fetch).toHaveBeenCalled();
    // Can do toHaveBeenCalledWith
    global.fetch.mockReset();
  });

  it('handles state updating', async () => {
    const wrapper = mount(
      <MockedProvider>
        <CreateItem />
      </MockedProvider>
    );
    wrapper
      .find('#title')
      .simulate('change', { target: { value: 'Testing', name: 'title' } });
    wrapper
      .find('#price')
      .simulate('change', { target: { value: 10000, name: 'price' } });
    wrapper.find('#description').simulate('change', {
      target: { value: 'This is great.', name: 'description' },
    });
    await wait();
    expect(wrapper.find('CreateItem').instance().state).toMatchObject({
      title: 'Testing',
      price: 10000,
      description: 'This is great.',
    });
  });

  it('creates an item when the form is submitted', async () => {
    const item = fakeItem();
    const { title, description, price } = item;
    const mocks = [
      {
        request: {
          query: CREATE_ITEM_MUTATION,
          variables: {
            title,
            description,
            image: '',
            largeImage: '',
            price,
          },
        },
        result: {
          data: {
            createItem: {
              ...item,
              id: 'abc123',
              __typename: 'Item',
            },
          },
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <CreateItem />
      </MockedProvider>
    );
    wrapper
      .find('#title')
      .simulate('change', { target: { value: title, name: 'title' } });
    wrapper
      .find('#price')
      .simulate('change', { target: { value: price, name: 'price' } });
    wrapper.find('#description').simulate('change', {
      target: { value: description, name: 'description' },
    });
    Router.router = { push: jest.fn() };
    wrapper.find('form').simulate('submit');
    await wait();
    expect(Router.router.push).toHaveBeenCalled();
    expect(Router.router.push).toHaveBeenCalledWith({
      pathname: '/item',
      query: { id: 'abc123' },
    });
  });
});
