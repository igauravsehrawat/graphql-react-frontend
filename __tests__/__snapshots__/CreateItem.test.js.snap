// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem /> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="image"
    >
      Image
    </label>
    <input
      id="image"
      name="image"
      onChange={[Function]}
      type="file"
    />
    <label
      htmlFor="title"
    >
      Title
    </label>
    <input
      id="title"
      name="title"
      onChange={[Function]}
      placeholder="Enter title"
      required={true}
      type="text"
      value=""
    />
    <label
      htmlFor="price"
    >
      Price
    </label>
    <input
      id="price"
      name="price"
      onChange={[Function]}
      placeholder="Enter price"
      required={true}
      type="number"
      value={0}
    />
    <label
      htmlFor="description"
    />
    <textarea
      id="description"
      name="description"
      onChange={[Function]}
      placeholder="Enter description"
      required={true}
      value=""
    />
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
